torchao/core/config.py

```diff
@@ -26,4 +33,226 @@ def _transform(
+
+    Args:
+        config: An instance of AOBaseConfig subclass
+
+    Returns:
+        Dict representation of the config
+    """
+    if not isinstance(config, AOBaseConfig):
+        raise TypeError(f"Expected AOBaseConfig instance, got {type(config)}")
+
+    # Use the existing JSON encoder but return the dict directly
+    return json.loads(json.dumps(config, cls=ConfigJSONEncoder))
+
+
+ALLOWED_AO_MODULES = {
+    "torchao.quantization",
+}
+
+
+def config_from_dict(data: Dict[str, Any]) -> AOBaseConfig:
+    """
+    Create an AOBaseConfig subclass instance from a dictionary.
+
+    Args:
+        data: Dictionary containing serialized config data
+
+    Returns:
+        An instance of the appropriate AOBaseConfig subclass
+
+    Raises:
+        VersionMismatchError: If the stored version doesn't match the class version
+        ValueError: If deserialization fails for other reasons
+    """
+    if not isinstance(data, dict):
+        raise TypeError(f"Expected dictionary, got {type(data)}")
+
+    if "_type" not in data or "_data" not in data:
+        raise ValueError("Input dictionary missing required '_type' or '_data' fields")
+
+    type_path = data["_type"]
+    stored_version = data.get("_version", 1)
+    obj_data = data["_data"]
+
+    # Handle torch.dtype
+    if type_path == "torch.dtype":
+        import torch
+
+        return getattr(torch, obj_data)
+    # Try to find the class in any of the allowed modules
+    cls = None
+    for module_path in ALLOWED_AO_MODULES:
+        try:
+            module = importlib.import_module(module_path)
+            cls = getattr(module, type_path)
+            break  # Found the class, exit the loop
+        except (ImportError, AttributeError):
+            continue  # Try the next module
+
+    # If we couldn't find the class in any allowed module, raise an error
+    if cls is None:
+        allowed_modules_str = ", ".join(ALLOWED_AO_MODULES)
+        raise ValueError(
+            f"Failed to find class {type_path} in any of the allowed modules: {allowed_modules_str}"
+        )
+
+    # Check version - require exact match
+    current_version = getattr(cls, "VERSION", 1)
+    if stored_version != current_version:
+        raise VersionMismatchError(type_path, stored_version, current_version)
+
+    # Handle the case where obj_data is not a dictionary
+    if not isinstance(obj_data, dict):
+        if issubclass(cls, enum.Enum):
+            # For enums, convert string to enum value
+            return getattr(cls, obj_data)
+        else:
+            # For other primitive types, create an instance with the value
+            try:
+                return cls(obj_data)
+            except:
+                return obj_data
+
+    # Process nested structures for dictionary obj_data
+    processed_data = {}
+    for key, value in obj_data.items():
+        if isinstance(value, dict) and "_type" in value and "_data" in value:
+            # Recursively handle nested configs
```

Comment:
Do we have any existing configs that have nested configs or fields with data structures? I feel our configs so far have been pretty simple?

Reply:
Vasiliy mentioned there might be one for float8 training but I could find it for config wrapping configs

We do have configs that wrap objects that need to be handled specially, the float inference configs have this where MMconfig is internal to it and needs special care

Reply:
> We do have configs that wrap objects that need to be handled specially, the float inference configs have this where MMconfig is internal to it and needs special care

I don't think MMConfig should be in the BC surface / serializable / etc, that seems like a code smell.

Reply:
I disagree it is a convenient  way for users to modulate mm behavior on a per instance level and unpacking this struct into three separate booleans doen't feel beneficial. 

That being said there are alot of examples of these nested objects, `Layouts`, `granularities`, etc.. for which this behavior is necessary